### ISSUE TITLE
issue while training using DDP

### ISSUE THREAD

**Issue by minwoo-yu** [SHI-Labs/NATTEN#166]
First of all, thank you for your great contributions to this community.

Unfortunately, I had a problem while training on a DDP with multiple GPUs :(
Training under a single GPU is well-performing, but when I tried training using 2 GPUs, an error occurred in the second GPU’s process. (I used NeighborhoodAttention2D and an error occurred in "attn = attn.softmax(dim=-1)")

An error occurred with the "NATTEN failure: cutlass error: Error Internal at: 102" messages.
As I searched, another user reported this error number and answered with a contiguous problem, but I think this one is not related to contiguous().

My running environment is organized as follows:
GPU: NVIDIA A6000 x2
Pytorch version: 2.4.1

**Comment by alihassanijr** [comment]
Thank you for your interest.

Could you share the version of NATTEN you're using, and kernel size, dilation, and other parameters related to NA?

**Comment by alihassanijr** [comment]
Is this FP16 or FP32?

Dim 180 is definitely misaligned for anything below FP32, so that might be the reason. But I'm very surprised an error isn't raised in the first place to prevent this. Looking into this.

**Comment by minwoo-yu** [comment]
I'm trying to training under FP32. 
Also, same errors are occurred when I change the dimension :(

**Comment by alihassanijr** [comment]
Okay that's concerning.

Could you please share a minimum steps/script to reproduce?

Also, if you're training in eager mode, it is also possible that some other operation is the issue. I'm assuming you've already tried running it with `CUDA_LAUNCH_BLOCKING=1`?

**Comment by AdityaKane2001** [comment]
@alihassanijr can this be related to the CUDAStream bug that we saw a while ago (which you actually fixed in 0.17)?

**Comment by alihassanijr** [comment]
I don't think so, and if it was fixed in 0.17 it shouldn't occur here since they're using 0.17.1.

**Comment by minwoo-yu** [comment]
I found out under which condition the error occurs!
This happens when the model is wrapped using distributed data parallel for training.

here is the demo:
[nattentest.py.zip](https://github.com/user-attachments/files/17091812/nattentest.py.zip)

**Comment by alihassanijr** [comment]
Okay I'm running this on my end, and I am able to reproduce the issue.

This is indeed unexpected and very strange. What I've found is that virtually all available kernels (naive, tiled, gemm, fna) hit the same error, even though some of them (fna) have a completely different interface. So this made me think it might be something that PyTorch expects extensions to do for distributed/multiprocess settings that we're not doing in NATTEN, and turns out I was right! ([Issue in the PT forum](https://discuss.pytorch.org/t/cuda-illegal-memory-access-for-cuda-extension-in-multiprocessing/198767/2)).

What is strange is that since the very first version of NATTEN back in April 2022 and to this day, I've probably ran in excess of 1000 experiments, all of which were done with DDP, and yet your issue was perfectly reproducible on one of the same machines I conducted those experiments over.

Anyways, the fix is very easy, and I've confirmed it fixes your script. I'll be opening a pull request soon and referencing this issue. 

If you're able to build NATTEN from source and confirm that it does indeed solve your original case, that would greatly help us. Otherwise we'll just send you an early wheel before pushing a new release over PyPI.

Thank you for bringing this to our attention, and for helping us get to the bottom of it!

**Comment by AdityaKane2001** [comment]
@alihassanijr 
If possible could you please run the same minimal code that @minwoo-yu shared with the following change?

```diff
def main(rank: int, ddp=True, n_gpus=2):
    if ddp:
        torch.distributed.init_process_group(
            backend="nccl",
            init_method="tcp://127.0.0.1:23456",
            world_size=n_gpus,
            rank=rank,
        )
    if ddp:
        model = DDP(Model().cuda(rank), device_ids=[rank])
+       torch.cuda.set_device(device)
        # model = DDP(make_model(2, "pixelshuffle", 1).cuda(rank), device_ids=[rank])
    else:
        model = Model().cuda(rank)
        # model = make_model(2, "pixelshuffle", 1).cuda(rank)
    for i in tqdm(range(10000), disable=(rank != 0)):
        input = torch.rand(2, 1, 96, 96, device=rank)
        output = model(input)
```


The reason, I believe, NATTEN didn't fail on any of the prior experiments is because `timm` always calls [`torch.cuda.set_device(device)`](https://github.com/huggingface/pytorch-image-models/blob/main/timm/utils/distributed.py#L171C9-L171C38) before returning the device object. As far as I understand, this essentially calls [`cudaSetDevice`](https://developer.download.nvidia.com/compute/DevZone/docs/html/C/doc/html/group__CUDART__DEVICE_g418c299b069c4803bfb7cab4943da383.html) under the hood, exactly what the optional guard calls on initialization.

**Comment by alihassanijr** [comment]
Wow good catch @AdityaKane2001 . I found another issue (#13) related where we had to recommend to users to definitely call `set_device` in their programs. 

I _think_ dlopening, which any torch extension would do, creates a subprocess or a new cuda context with device defaulting to 0, and that's why we need the device guard. But I'm unsure how `set_device` would solve that.

Anyway, it's great that we have more information about it!

**Comment by AdityaKane2001** [comment]
I don't have code that I can point to, but intuitively `set_device` (or the `OptionalCUDAGuard`) would cause all tensors (weights and activations) to be allocated and manipulated on a particular device (consequently, setting a unique context and later a unique stream). This will sidestep any illegal access errors. 

Again, I want to stress that this is my intuition without any solid backing. I will try to find evidence in code to support this.

**Comment by alihassanijr** [comment]
It's most definitely not tensor allocation. libnatten purposefully never does `cudaMalloc` on its own, because we want to use the PT caching allocator, so everything is defined as torch tensors. With the exception of scratch space for FNA (which isn't the use case here), all other torch tensors are defined in the python API and passed down.

Tensors passed down are always on the correct device, so as long as the torch api picks up the correct stream and sets the correct cuda context, there shouldn't be any issues.
